Store/Receivers: Calculating chunk hashes on stores/receivers

pedro-stanaka opened the pull request:
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

Related to #5682 

## Changes

* Change multiple implementations of `Series()` in store to pre-calculate hashes for faster deduplication on querier:
  - TSDB
  - Prometheus
  - Block storage 

## Verification

<!-- How you tested it? How do you know it works? -->

yeya24 commented on pkg/store/bucket.go:
```diff
@@ -885,7 +886,7 @@ func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk, aggrs []storepb.Ag
 			return err
 		}
-		out.Raw = &storepb.Chunk{Type: storepb.Chunk_XOR, Data: b}
+		out.Raw = &storepb.Chunk{Type: storepb.Chunk_XOR, Data: b, Hash: xxhash.Sum64(b)}
```
Let's add a flag to this https://github.com/thanos-io/thanos/blob/main/pkg/store/storepb/rpc.proto#L88?
Since store gateway is also used by other projects like Cortex, it is better to control this behavior instead of enabling it by default.

yeya24 commented on pkg/store/bucket.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"context"
 	"encoding/binary"
 	"fmt"
+	"github.com/cespare/xxhash"
```
Please group the import

pedro-stanaka replied:
thx, was a preliminary version. fixed

bwplotka commented on pkg/store/storepb/rpc.proto:
```diff
@@ -122,6 +122,9 @@ message SeriesRequest {
   // shard of blocks instead of entire blocks.
   ShardInfo shard_info = 13;
+
+  // Control whether the store should calculate the checksum/hash of chunks
```
```suggestion
  // Control whether the store should calculate the checksum/hash of chunks,
```

pedro-stanaka replied:
Out of curiosity, why the comma here?

GiedriusS replied:
I think it should be `.` :smile:

pedro-stanaka replied:
Put a dot `.` now. Makes more sense. Thx

bwplotka commented on pkg/store/storepb/rpc.proto:
```diff
@@ -122,6 +122,9 @@ message SeriesRequest {
   ShardInfo shard_info = 13;
+
+  // Control whether the store should calculate the checksum/hash of chunks
+  bool calculate_chunk_checksums = 14;
```
Should we remove this and control that in the implementations (each server option for all requests). I don't see a need for request to ask for it explicitly 🤔

fpetkovski replied:
I think it's because of https://github.com/thanos-io/thanos/pull/5703#discussion_r974439342

yeya24 replied:
On the Thanos side I think we can always enable it. But Cortex also relies on the store gateway and the store API and calculating hash on the cortex side is unnecessary for now

pedro-stanaka replied:
Please let me know what is best here. :)

fpetkovski replied:
I think it's fine to have it as a command-line argument instead of adding it to the request itself. That's what @bwplotka had in mind I believe.

yeya24 replied:
Umm yeah, sounds like a flag on store gateway is better

pedro-stanaka replied:
Okay, I will add flag. But then we need flag for:

* Receiver (MultiTSDB -> TSDB)
* Ruler (TSDB)
* Store Gateway (Bucket Store)
* Sidecar (Prometheus Store)

Is that right?

yeya24 replied:
If we don't need to change the protobuf file, then seems it is also okay to not add any flags.
Since Cortex only shares the store gateway with Thanos, so as long as the `blockSeries` method exposes a boolean param to control not calculate hash then it is fine. On the thanos side, we can make it always `true`.

pedro-stanaka replied:
If I got this right, we would need still to let them pass an option here: https://github.com/cortexproject/cortex/blob/d4fceb1172fff9ac332e525086d07f2ef182aefc/pkg/storegateway/bucket_stores.go#L484, right? Should I create a field on the bucket store to set wether or not to calculate hashes? Then, they could write a `BucketStoreOption` func that allows them to set this field. Let me know if this is okay.

yeya24 replied:
Sorry for the late reply. For me I think it is fine to add it at `blockSeries` to control that option. Do you think it works?

pedro-stanaka replied:
I did that, I just thought that since that is a private function of the package they were not able to call it themselves.

yeya24 replied:
Ah you are right sorry. So we need either:
1. an option at BucketStore to turn on/off chunk hashing
2. modify series request

pedro-stanaka replied:
Can you please recheck @yeya24 , I went with my idea to create a `BucketStoreOption` and enabled it by default on thanos side.

yeya24 commented on pkg/store/bucket.go:
```diff
@@ -945,12 +949,19 @@ func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk, aggrs []storepb.Ag
+	if !doHash {
+		return 0
+	}
+	return xxhash.Sum64(b)
```
We could probably have a sync pool for `xxhash` and reuse `xxhash.New()`.

pedro-stanaka replied:
Can you check if I used it correctly? TBH, I have never used it before, but I checked examples in the codebase. Thx for the review.

pedro-stanaka commented on pkg/store/bucket.go:
```diff
@@ -872,20 +876,29 @@ func blockSeries(
+		New: func() interface{} {
+			return xxhash.New()
+		},
+	}
```
I have adopted the way that Cortex uses, so I can reuse across the package. Thanks for the example.

yeya24 commented on pkg/store/bucket.go:
```diff
@@ -945,12 +950,25 @@ func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk, aggrs []storepb.Ag
 
+func hashChunk(b []byte, doHash bool) uint64 {
+	hasher := hashPool.Get().(hash.Hash64)
+	defer hashPool.Put(hasher)
```
Probably a better way to go is to extract these 2 lines at the beginning of `populateChunk` function. This can reuse   `hash.Hash64` and avoid some lock contention of the pool.

yeya24 reviewed (approved):
LGTM

pedro-stanaka commented on pkg/store/prometheus.go:
```diff
@@ -402,7 +411,12 @@ func (p *PrometheusStore) handleStreamedPrometheusResponse(
 			for i, chk := range series.Chunks {
+				hash := uint64(0)
+				if calculateChecksums {
+					hash = xxhash.Sum64(chk.Data)
```
Done.

yeya24 reviewed (approved):
LGTM

pedro-stanaka commented:
> @pedro-stanaka can you push an empty commit please to retrigger the CI? I think something happened to Github Actions and they didn't run after your last push.

Hey Douglas. Thanks for taking a look at this. I rebased and fixed some errors regarding new code from main, should be fine now.

fpetkovski reviewed (approved):
This looks good. Should we merge it?

yeya24 reviewed (approved):
LGTM

GiedriusS reviewed (commented):
Now we need to start using this [here](https://github.com/thanos-io/thanos/blob/cc4e60724f2ed3c5025fdf3fb93f6a08a4ce528d/pkg/store/proxy_heap.go#L134) ;P
